library/ssl_srv.c

```diff
@@ -4003,13 +4003,28 @@ static int ssl_parse_certificate_verify( mbedtls_ssl_context *ssl )
 
     MBEDTLS_SSL_DEBUG_MSG( 2, ( "=> parse certificate verify" ) );
 
-    if( !mbedtls_ssl_ciphersuite_cert_req_allowed( ciphersuite_info ) ||
-        ssl->session_negotiate->peer_cert == NULL )
+    if( !mbedtls_ssl_ciphersuite_cert_req_allowed( ciphersuite_info ) )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 2, ( "<= skip parse certificate verify" ) );
+        ssl->state++;
+        return( 0 );
+    }
+
+#if defined(MBEDTLS_SSL_KEEP_PEER_CERTIFICATE)
+    if( ssl->session_negotiate->peer_cert == NULL )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 2, ( "<= skip parse certificate verify" ) );
+        ssl->state++;
+        return( 0 );
+    }
+#else /* MBEDTLS_SSL_KEEP_PEER_CERTIFICATE */
```

Comment:
Minor: I think it would be clearer to isolate the decision-making (are we going to skip?) from the execution (inc state and return early), as currently having (visually) 3 `state++` in the function hinders readability. Could be done later (as part of MPS prepartion with a `coordinate` function?)

Reply:
Yes, I fully agree, and #1634 does that, see https://github.com/ARMmbed/mbedtls/blob/62594db239cf337ed5498f1b9e11796dfd679a76/library/ssl_srv.c#L4259. Given the time pressure, I'd prefer to not change anything in this PR. Are you ok with that?

Reply:
Sure! As I said, most of my remarks don't have to result in code changes now - except the two things marked "major" (and even perhaps only one of them, if you can convince me about the one in `ssl_client2.c`).